plugins/bcm57xx/fu-bcm57xx-recovery-device.c

```diff
@@ -0,0 +1,753 @@
+					     error);
+	if (locker2 == NULL)
+		return FALSE;
+
+	/* get NVRAM version */
+	if (!fu_bcm57xx_recovery_device_nvram_read (self, BCM_NVRAM_STAGE1_BASE + BCM_NVRAM_STAGE1_VERSION,
+					   &fwversion, 1, error))
+		return FALSE;
+	if (fwversion != 0x0) {
+		g_autofree gchar *fwversion_str = NULL;
+
+		/* this is only set on the non-free firmware */
+		fwversion_str = fu_common_version_from_uint32 (GUINT32_FROM_BE(fwversion),
+							       FWUPD_VERSION_FORMAT_TRIPLET);
+		fu_device_set_version_format (device, FWUPD_VERSION_FORMAT_TRIPLET);
+		fu_device_set_version (device, fwversion_str);
+		fu_device_set_version_raw (device, fwversion);
+		fu_device_set_branch (device, BCM_FW_BRANCH_OSS_FIRMWARE);
+	} else {
+		guint32 bufver[4] = { 0x0 };
+		guint32 veraddr = 0;
+		g_autoptr(Bcm57xxVeritem) veritem = NULL;
+
+		/* fall back to the string, e.g. '5719-v1.43' */
+		if (!fu_bcm57xx_recovery_device_nvram_read (self,
+							    BCM_NVRAM_STAGE1_BASE + BCM_NVRAM_STAGE1_VERADDR,
+							    &veraddr, 1, error))
+			return FALSE;
+		veraddr = GUINT32_FROM_BE(veraddr);
+		if (veraddr > BCM_PHYS_ADDR_DEFAULT)
+			veraddr -= BCM_PHYS_ADDR_DEFAULT;
+		if (!fu_bcm57xx_recovery_device_nvram_read (self,
+						   BCM_NVRAM_STAGE1_BASE + veraddr,
+						   bufver, 4, error))
+			return FALSE;
+		veritem = fu_bcm57xx_veritem_new ((guint8 *) bufver, sizeof(bufver));
+		if (veritem != NULL) {
+			fu_device_set_version (device, veritem->version);
+			fu_device_set_branch (device, veritem->branch);
+			fu_device_set_version_format (device, veritem->verfmt);
+		}
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_open (FuDevice *device, GError **error)
+{
+	FuBcm57xxRecoveryDevice *self = FU_BCM57XX_RECOVERY_DEVICE (device);
+#ifdef HAVE_MMAN_H
+	FuUdevDevice *udev_device = FU_UDEV_DEVICE (device);
+	const gchar *sysfs_path = fu_udev_device_get_sysfs_path (udev_device);
+#endif
+
+#ifdef RUNNING_ON_VALGRIND
+	/* this can't work */
+	if (RUNNING_ON_VALGRIND) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_NOT_SUPPORTED,
+				     "cannot mmap'ing BARs when using valgrind");
+		return FALSE;
+	}
+#endif
+
+#ifdef HAVE_MMAN_H
+	/* map BARs */
+	for (guint i = 0; i < FU_BCM57XX_BAR_MAX; i++) {
+		int memfd;
+		struct stat st;
+		g_autofree gchar *fn = NULL;
+		g_autofree gchar *resfn = NULL;
+
+		/* open 64 bit resource */
+		resfn = g_strdup_printf ("resource%u", i * 2);
+		fn = g_build_filename (sysfs_path, resfn, NULL);
+		memfd = open (fn, O_RDWR | O_SYNC);
+		if (memfd < 0) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_NOT_FOUND,
+				     "error opening %s", fn);
+			return FALSE;
+		}
+		if (fstat (memfd, &st) < 0) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_NOT_SUPPORTED,
+				     "could not stat %s", fn);
+			close (memfd);
+			return FALSE;
+		}
+
+		/* mmap */
+		g_debug ("mapping BAR[%u] %s for 0x%x bytes", i, fn, (guint) st.st_size);
```

Comment:
this should probably be hidden behind a plugin verbose flag

Reply:
Agree, fixed.